File: context.go
package microscope

import (
	"context"
	"strings"
)

type ctxKey string

const keyBatchID ctxKey = "microscope_batch_id"
const keySkipTrace ctxKey = "microscope_skip_trace"

// WithBatchID attaches the current request batch to the context.
func WithBatchID(ctx context.Context, batchID string) context.Context {
	return context.WithValue(ctx, keyBatchID, batchID)
}

// BatchIDFromContext returns the batch ID for the current request, if any.
func BatchIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keyBatchID).(string)
	return v
}

// WithoutTrace marks ctx so microscope query tracing is skipped (internal DB ops).
func WithoutTrace(ctx context.Context) context.Context {
	return context.WithValue(ctx, keySkipTrace, true)
}

func traceSkipped(ctx context.Context) bool {
	v, _ := ctx.Value(keySkipTrace).(bool)
	return v
}

// IsMicroscopePath reports whether path belongs to the microscope dashboard/API.
func IsMicroscopePath(path, prefix string) bool {
	if prefix == "" {
		prefix = "/microscope"
	}
	return path == prefix || strings.HasPrefix(path, prefix+"/")
}

File: adaptor/python/microscope/context.py
from __future__ import annotations

from contextvars import ContextVar

_batch_id: ContextVar[str] = ContextVar("microscope_batch_id", default="")
_skip_trace: ContextVar[bool] = ContextVar("microscope_skip_trace", default=False)


def with_batch_id(batch_id: str) -> None:
    _batch_id.set(batch_id)


def batch_id_from_context() -> str:
    return _batch_id.get()


def without_trace() -> None:
    _skip_trace.set(True)


def trace_skipped() -> bool:
    return _skip_trace.get()


def is_microscope_path(path: str, prefix: str = "/microscope") -> bool:
    if not prefix:
        prefix = "/microscope"
    return path == prefix or path.startswith(f"{prefix}/")
